hugo/content/docs/ast-types.md

```diff
@@ -3,4 +3,340 @@ title: "AST Type Inference"
+```
+X: 
+    {infer A} 'A' name=ID 
+  | {infer B} 'B' name=ID count=INT;
+
+// is equivalent to:
+X: A | B;
+A: 'A' name=ID;
+B: 'B' name=ID count=INT;
+
+// generates:
+type X = A | B;
+
+interface A extends AstNode {
+    name: string
+}
+
+interface B extends AstNode {
+    name: string
+    count: number
+}
+```
+
+### Assigned actions
+Actions can also be used to control the structure of AST node types. This is a more advanced topic, so we recommend getting familiar with the rest of the documentation before diving into this section.
+
+Let's consider two different grammars derived from the [Arithmetics example](https://github.com/langium/langium/blob/main/examples/arithmetics/src/language-server/arithmetics.langium). These grammars are designed to parse a document containing a single definition comprised of a name and an expression assignment, with an expression being an indefinite length of additions or a numerical value.
+
+The first one does not use assigned actions:
+```
+Definition: 
+    'def' name=ID ':' expr=Expression;
+
+Expression:
+    '(' Addition ')' | value=NUMBER;
+
+Addition infers Expression:
+    left=Expression ('+' right=Expression)*;
+```
+When parsing a document containing `def x: (1 + 2) + 3`, this is the shape of the AST:
+{{<mermaid>}}
+graph TD;
+expr((expr)) --> left((left))
+expr --> right((right))
+left --> left_left((left))
+left --> left_right((right))
+right --> right_right((right))
+left_left --> left_left_v{1}
+left_right --> left_right_{2}
+right_right --> right_right_v{3}
+{{</mermaid>}}
+We can see that the nested `right -> right` property in the tree is counter-intuitive and we would like to remove one level of nesting from the tree. 
+
+This can be done by reworking the grammar and adding an assigned action:
+```
+Definition: 
+    'def' name=ID ':' expr=Addition ';';
+
+Expression:
+    '(' Addition ')' | value=NUMBER;
+
+Addition infers Expression:
+    Expression ({infer Addition.left=current} '+' right=Expression)*;
+```
+Parsing the same document now leads to this AST:
+{{<mermaid>}}
+graph TD;
+expr((expr)) --> left((left))
+expr --> right((right))
+left --> left_left((left))
+left --> left_right((right))
+right --> right_v{3}
+left_left --> left_left_v{1}
+left_right --> left_right_{2}
+{{</mermaid>}}
+
+## Declared Types
+Because type inference takes into account every entity of a parser rule, even the smallest changes can update the inferred types. This can lead to an unexpected type system and incorrect behavior of services that depend on it. To minimize the risk of introducing breaking changes when modifying the grammar, we recommend to use *declared types*. This is especially true for more mature and complex languages, where a stable type system is key and breaking changes introduced by inferred types can be hard to detect. Declared types allow the user to **fix** the type of the parser rules and rely on the power of validation errors to detect breaking changes.
```

Comment:
```suggestion
Because type inference takes into account every entity of a parser rule, even the smallest changes can update the inferred types. This can lead to an unexpected meta model and incorrect behavior of services that depend on it. To minimize the risk of introducing breaking changes when modifying the grammar, we recommend to use *declared types*. This is especially true for more mature and complex languages, where a stable type system is key and breaking changes introduced by inferred types can be hard to detect. Declared types allow the user to **fix** the type of the parser rules and rely on validation errors to detect breaking changes.
```